bot/exts/utilities/color.py

```diff
@@ -0,0 +1,229 @@
+
+    @colour.command()
+    async def hex(self, ctx: commands.Context, hex_code: str) -> None:
+        """Command to create an embed from a HEX input."""
+        if "#" not in hex_code:
+            hex_code = f"#{hex_code}"
+        hex_tuple = ImageColor.getrgb(hex_code)
+        await self.send_colour_response(ctx, hex_tuple)
+
+    @colour.command()
+    async def name(self, ctx: commands.Context, user_colour_name: str) -> None:
+        """Command to create an embed from a name input."""
+        hex_colour = self.match_colour_name(user_colour_name)
+        hex_tuple = ImageColor.getrgb(hex_colour)
+        await self.send_colour_response(ctx, hex_tuple)
+
+    @colour.command()
+    async def random(self, ctx: commands.Context) -> None:
+        """Command to create an embed from a randomly chosen colour from the reference file."""
+        hex_colour = random.choice(list(self.colour_mapping.values()))
+        hex_tuple = ImageColor.getrgb(f"#{hex_colour}")
+        await self.send_colour_response(ctx, hex_tuple)
+
+    def get_colour_conversions(self, rgb: tuple[int, int, int]) -> dict[str, str]:
+        """Create a dictionary mapping of colour types and their values."""
+        return {
+            "RGB": rgb,
+            "HSV": self._rgb_to_hsv(rgb),
+            "HSL": self._rgb_to_hsl(rgb),
+            "CMYK": self._rgb_to_cmyk(rgb),
+            "Hex": self._rgb_to_hex(rgb),
+            "Name": self._rgb_to_name(rgb)
+        }
+
+    @staticmethod
+    def _rgb_to_hsv(rgb: tuple[int, int, int]) -> tuple[int, int, int]:
+        """Convert RGB values to HSV values."""
+        rgb_list = [val / 255 for val in rgb]
+        h, s, v = colorsys.rgb_to_hsv(*rgb_list)
+        hsv = (round(h * 360), round(s * 100), round(v * 100))
+        return hsv
+
+    @staticmethod
+    def _rgb_to_hsl(rgb: tuple[int, int, int]) -> tuple[int, int, int]:
+        """Convert RGB values to HSL values."""
+        rgb_list = [val / 255.0 for val in rgb]
+        h, l, s = colorsys.rgb_to_hls(*rgb_list)
+        hsl = (round(h * 360), round(s * 100), round(l * 100))
+        return hsl
+
+    @staticmethod
+    def _rgb_to_cmyk(rgb: tuple[int, int, int, int]) -> tuple[int, int, int, int]:
+        """Convert RGB values to CMYK values."""
+        rgb_list = [val / 255.0 for val in rgb]
+        if not any(rgb_list):
+            return 0, 0, 0, 100
+        k = 1 - max(val for val in rgb_list)
+        c = round((1 - rgb_list[0] - k) * 100 / (1 - k))
+        m = round((1 - rgb_list[1] - k) * 100 / (1 - k))
+        y = round((1 - rgb_list[2] - k) * 100 / (1 - k))
+        cmyk = (c, m, y, round(k * 100))
+        return cmyk
+
+    @staticmethod
+    def _rgb_to_hex(rgb: tuple[int, int, int]) -> str:
+        """Convert RGB values to HEX code."""
+        hex_ = ''.join([hex(val)[2:].zfill(2) for val in rgb])
+        hex_code = f"#{hex_}".upper()
+        return hex_code
+
+    def _rgb_to_name(self, rgb: tuple[int, int, int]) -> str:
```

Comment:
Perhaps the error percentage of closeness could be listed as well?

I'm not even sure that rapidfuzz is needed here, or is the best tool for the job.

There's difflib, which is part of the stdlib. 

In addition, there is a great deal of significance with the naming of these, and how they're matched.

As a quick review, because of how rgb and hex work, fff is very different from 000...
but run `rapidfuzz.process.extractOne('f07070',['fefefe','010101'])` and the result is 010101.

The problem here is rapidfuzz thinks they're close together, but they aren't. The significance of the 1st, 3rd, and 5th characters is much more important than the 2nd, 4th, and 6th, characters.


 
A possible solution for this is two things, a function that caches the entire list of keys, and turns them all into integer values.

Next, a binary search of that list to find the value that most closely matches the provided integer value of the colour.

There may be a better solution than the one I just discovered, so please explore some ideas before implementing it.

Reply:
One way would be to have the colours as r b g integer tuples and find the one with the smallest euclidean distance using `math.dist`
 ```python
colours = [(30, 30, 100), (25, 40, 25), (25, 25, 25)]
target_colour = (30, 30, 40)
closest = min(colours, key=lambda colour: math.dist(colour, target_colour))
# closest is (30, 30, 30)
```
Comparing colours is actually quite an interesting topic, and whilst this solution isn't perfect due to the way RGB works, I think it would be good enough for our purposes.

Reply:
So for either of these solutions, I need the `ryanzec.json` list to have an rgb value: color name key-value pair?

Reply:
Not implementing in the base PR as mentioned here: https://discord.com/channels/267624335836053506/635950537262759947/907684453806112818